corehq/form_processor/reprocess.py

```diff
@@ -0,0 +1,206 @@
+from collections import defaultdict, namedtuple
+
+from couchdbkit import ResourceNotFound
+
+from casexml.apps.case.signals import case_post_save
+from corehq.blobs.mixin import bulk_atomic_blobs
+from corehq.form_processor.backends.sql.dbaccessors import FormAccessorSQL, CaseAccessorSQL, LedgerAccessorSQL
+from corehq.form_processor.change_publishers import publish_form_saved
+from corehq.form_processor.exceptions import XFormNotFound
+from corehq.form_processor.interfaces.processor import FormProcessorInterface
+from corehq.form_processor.models import XFormInstanceSQL, CaseTransaction, LedgerTransaction, FormReprocessRebuild
+from corehq.form_processor.submission_post import SubmissionPost
+from corehq.form_processor.utils.general import should_use_sql_backend
+from corehq.sql_db.util import get_db_alias_for_partitioned_doc
+from couchforms.models import XFormInstance
+
+import logging
+
+
+ReprocessingResult = namedtuple('ReprocessingResult', 'form cases ledgers')
+
+logger = logging.getLogger('reprocess')
+
+
+def reprocess_unfinished_stub(stub, save=True):
+    if stub.saved:
+        # ignore for now
+        logger.info("Ignoring 'saved' stub: %s", stub.xform_id)
+        return
+
+    if not should_use_sql_backend(stub.domain):
+        # ignore for couch domains
+        logger.info('Removing stub from non SQL domain: %s', stub.xform_id)
+        save and stub.delete()
+        return
+
+    form_id = stub.xform_id
+    try:
+        form = FormAccessorSQL.get_form(form_id)
+    except XFormNotFound:
+        # form doesn't exist which means the failure probably happend during saving so
+        # let mobile handle re-submitting it
+        logger.error('Form not found: %s', form_id)
+        save and stub.delete()
+        return
+
+    assert form.is_normal
+    result = _reprocess_form(form)
+    save and stub.delete()
+    return result
+
+
+def reprocess_xform_error(form):
+    """
+    Attempt to re-process an error form. This was created specifically to address
+    the issue of out of order forms and child cases (form creates child case before
+    parent case has been created).
+
+    See http://manage.dimagi.com/default.asp?250459
+    :param form_id: ID of the error form to process
+    """
+    if not form:
+        raise Exception('Form with ID {} not found'.format(form.form_id))
+
+    if not form.is_error:
+        raise Exception('Form was not an error form: {}={}'.format(form.form_id, form.doc_type))
```

Comment:
Would it make sense to raise a custom subclass of `Exception` so it can be reliably caught without possibly masking bugs?

Also noticed several places below where `Exception` is raised.